spec.html

```diff
@@ -9473,6 +9466,19 @@ <h1>IntegerIndexedObjectCreate ( _prototype_, _internalSlotsList_ )</h1>
         </emu-alg>
       </emu-clause>
 
+      <emu-clause id="sec-integerindexedelementget" aoid="IntegerIndexedElementGet">
+        <h1>IsIntegerIndexValid ( _O_, _index_ )</h1>
+        <p>The abstract operation IsIntegerIndexValid with arguments _O_ and _index_ performs the following steps:</p>
+        <emu-alg>
+          1. Assert: Type(_index_) is Number.
+          1. Assert: _O_ is an Object that has an [[ArrayLength]] internal slot.
```

Comment:
is this a shorthand for:
```suggestion
          1. Assert: _O_ is an Integer-Indexed exotic object.
```
?

if so, can we use that definition?

Reply:
It's a more relaxed check than asserting something is an Integer-Indexed exotic object, which I take to mean it would assert that the object has *all* the internal slots as well as the Integer-Indexed internal methods.

The spec has a convention to only do the brand check that it needs, and I think that flexibility is nice for writing future specs. I prefer to keep it to just the [[ArrayLength]] assert.

Reply:
To me it seems more brittle to future changes. Is there anything that would have that internal slot and not have all the others?

Additionally, the auto linking to the definition would likely be helpful.

Reply:
I don't understand the brittleness concern. Isn't it more brittle (less flexible) to assert that it is an IIEO? Do we have precedence for that kind of assertion over checking for specific slots?

Reply:
Some precendent, yes. E.g., checking whether something is:
- a Bound Function exotic object (vs having `[[BoundTargetFunction]]`, `[[BoundThis]]`, `[[BoundArguments]]` slots),
- a Proxy [exotic] object (vs having `[[ProxyHandler]]`, `[[ProxyTarget]]` slots),
- an AsyncGenerator instance (vs having `[[AsyncGenerator{State,Context,Queue}]]` slots),
- an ECMAScript function [object] (vs having the slots of Table 29)

Reply:
(However, checking for the presence/absence of particular slots is about 5 times more popular.)

Reply:
I read these assertions like type annotations; imo we should be checking for a nominal type here, not a structural one.

Reply:
Good to know there's precedence for both.

@ljharb and @jmdyck, do you think all exotic object checking should be implemented via "is a Foo exotic object"?

In general I like the explicit asserting of slots approach better, because all "exotic object" means is that it has special behavior for one of the essential internal methods. That it might have a set of distinctive internal slots is probably an implementation necessity for the exotic behavior, not an intrinsic property of being an exotic object. What do you think?

Reply:
I think that if the special behavior appears on multiple kinds of exotic object, it's definitely appropriate to assert on that special behavior alone. However, if there's only one kind of thing with that behavior, it's clearer to assert that you have a special kind of thing everywhere, so that if in the future new things are added with the behavior, the refactorings of all relevant usage sites are forced and explicit.

Reply:
> However, if there's only one kind of thing with that behavior, it's clearer to assert that you have a special kind of thing everywhere, so that if in the future new things are added with the behavior, the refactorings of all relevant usage sites are forced and explicit.

I see editorial arguments both ways. Doing slot checks (and structural checking) is more in line with the ethos of having generic abstract operations that "just work" for future things, i.e. better compositionality. Doing "is Foo exotic object" checks is more in line with the ethos of narrower operations, which may guard against undesirable genericity.

For this particular PR, TypedArray operations are most definitely not generic, so I'm fine with changing the asserts in that section to "is an IIEO".

Reply:
Happy to continue considering case by case, for sure.